openmeter/billing/service/lineservice/usagebasedline_test.go

```diff
@@ -0,0 +1,1366 @@
+				},
+			},
+		})
+	})
+
+	t.Run("usage present, mid line", func(t *testing.T) {
+		runUBPTest(t, ubpCalculationTestCase{
+			price: plan.NewPriceFrom(plan.UnitPrice{
+				Amount: alpacadecimal.NewFromFloat(100),
+			}),
+			lineMode: midPeriodSplitLineMode,
+			usage: featureUsageResponse{
+				LinePeriodQty: alpacadecimal.NewFromFloat(10),
+			},
+			expect: newDetailedLinesInput{
+				{
+					Name:                   "feature: usage in period",
+					Amount:                 alpacadecimal.NewFromFloat(100),
+					Quantity:               alpacadecimal.NewFromFloat(10),
+					ChildUniqueReferenceID: UnitPriceUsageChildUniqueReferenceID,
+					PaymentTerm:            plan.InArrearsPaymentTerm,
+				},
+			},
+		})
+	})
+
+	// Max spend is always honored
+	t.Run("usage present, max spend set, but not hit", func(t *testing.T) {
+		runUBPTest(t, ubpCalculationTestCase{
+			price: plan.NewPriceFrom(plan.UnitPrice{
+				Amount:        alpacadecimal.NewFromFloat(10),
+				MaximumAmount: lo.ToPtr(alpacadecimal.NewFromFloat(100)),
+			}),
+			lineMode: midPeriodSplitLineMode,
+			usage: featureUsageResponse{
+				LinePeriodQty: alpacadecimal.NewFromFloat(10),
+			},
+			expect: newDetailedLinesInput{
+				{
+					Name:                   "feature: usage in period",
+					Amount:                 alpacadecimal.NewFromFloat(10),
+					Quantity:               alpacadecimal.NewFromFloat(10),
+					ChildUniqueReferenceID: UnitPriceUsageChildUniqueReferenceID,
+					PaymentTerm:            plan.InArrearsPaymentTerm,
+				},
+			},
+		})
+	})
+
+	t.Run("usage present, max spend set, but not hit", func(t *testing.T) {
+		runUBPTest(t, ubpCalculationTestCase{
+			price: plan.NewPriceFrom(plan.UnitPrice{
+				Amount:        alpacadecimal.NewFromFloat(10),
+				MaximumAmount: lo.ToPtr(alpacadecimal.NewFromFloat(100)),
+			}),
+			lineMode: midPeriodSplitLineMode,
+			usage: featureUsageResponse{
+				LinePeriodQty:    alpacadecimal.NewFromFloat(5),
+				PreLinePeriodQty: alpacadecimal.NewFromFloat(7),
+			},
+			expect: newDetailedLinesInput{
+				{
+					Name:                   "feature: usage in period",
+					Amount:                 alpacadecimal.NewFromFloat(10),
+					Quantity:               alpacadecimal.NewFromFloat(5),
+					ChildUniqueReferenceID: UnitPriceUsageChildUniqueReferenceID,
+					PaymentTerm:            plan.InArrearsPaymentTerm,
+					Discounts: []billingentity.LineDiscount{
+						{
+							Description: lo.ToPtr("Maximum spend discount for charges over 100"),
+							Amount:      alpacadecimal.NewFromFloat(20),
+							Type:        lo.ToPtr(billingentity.MaximumSpendLineDiscountType),
+							Source:      billingentity.CalculatedLineDiscountSource,
+						},
+					},
+				},
+			},
+		})
+	})
+}
+
+func TestTieredGraduatedCalculation(t *testing.T) {
+	testTiers := []plan.PriceTier{
+		{
+			UpToAmount: lo.ToPtr(alpacadecimal.NewFromFloat(5)),
+			FlatPrice: &plan.PriceTierFlatPrice{
+				// 20/unit
+				Amount: alpacadecimal.NewFromFloat(100),
```

Comment:
Tiered pricing means units prices on amount see:
https://openmeter.io/blog/tiered-pricing-model#example-comparison

So the unit price will be $100 for for the first 5 items not $20